File: leetcode/145_Binary_Tree_Postorder_Traversal.cpp
#include<bits/stdc++.h>
using namespace std;
struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
  };
 
class Solution {
public:
    void post(TreeNode* root, vector<int>& res){
        if(root==nullptr)return ;
        post(root->left,res);
        post(root->right,res);
        res.push_back(root->val);
         
    }
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int>res;
        post(root, res);
        return res;
    }
};
int main(){
    Solution s;
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(3);
    vector<int> result = s.postorderTraversal(root);
    cout<<"Postorder Traversal: ";
    for(int val : result){
        cout<<val<<" ";
    }
    cout<<endl;
    return 0;
}

File: leetcode/145_Binary_Tree_Postorder_Traversal.java
import java.util.ArrayList;

 public class TreeNode {
     int val;
     TreeNode left;
     TreeNode right;
     TreeNode() {}
     TreeNode(int val) { this.val = val; }
     TreeNode(int val, TreeNode left, TreeNode right) {
         this.val = val;
         this.left = left;
         this.right = right;
     }
 }
 
class Solution {
    ArrayList<Integer> post= new ArrayList<>();
    public void BST(TreeNode root){
        if(root==null)return;
        BST(root.left);
        BST(root.right);
        post.add(root.val);
    }
    public List<Integer> postorderTraversal(TreeNode root) {
        BST(root);
        return post;
    }
    public List<Integer> postorderTraversal2(TreeNode root) {
        ArrayList<Integer> post= new ArrayList<>();
        if(root==null)return post;
        Stack<TreeNode> stack= new Stack<>();
        stack.push(root);
        while(!stack.isEmpty()){
            TreeNode node= stack.pop();
            post.add(node.val);
            if(node.left!=null)stack.push(node.left);
            if(node.right!=null)stack.push(node.right);
        }
        Collections.reverse(post);
        return post;
    }
}
public class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root == null) return res;
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            res.add(node.val);
            if (node.left != null) stack.push(node.left);
            if (node.right != null) stack.push(node.right);
        }
        Collections.reverse(res);
        return res;
    }
}